src/main/java/bisq/core/dao/node/consensus/TxInputController.java

```diff
@@ -20,43 +20,46 @@
 import bisq.core.dao.blockchain.ReadableBsqBlockChain;
 import bisq.core.dao.blockchain.WritableBsqBlockChain;
 import bisq.core.dao.blockchain.vo.SpentInfo;
-import bisq.core.dao.blockchain.vo.Tx;
 import bisq.core.dao.blockchain.vo.TxInput;
-import bisq.core.dao.blockchain.vo.TxOutput;
+import bisq.core.dao.blockchain.vo.TxOutputType;
 
 import javax.inject.Inject;
 
-import java.util.Optional;
-
 import lombok.extern.slf4j.Slf4j;
 
 /**
- * Provide spendable TxOutput and apply state change.
+ * Processes TxInput and add input value to available balance if the input is a valid BSQ input.
  */
-
 @Slf4j
 public class TxInputController {
 
-    private final WritableBsqBlockChain writableBsqBlockChain;
     private final ReadableBsqBlockChain readableBsqBlockChain;
 
     @Inject
-    public TxInputController(WritableBsqBlockChain writableBsqBlockChain, ReadableBsqBlockChain readableBsqBlockChain) {
-        this.writableBsqBlockChain = writableBsqBlockChain;
+    public TxInputController(ReadableBsqBlockChain readableBsqBlockChain) {
         this.readableBsqBlockChain = readableBsqBlockChain;
     }
 
-    Optional<TxOutput> getOptionalSpendableTxOutput(TxInput input) {
-        // TODO check if Tuple indexes of inputs outputs are not messed up...
-        // Get spendable BSQ output for txIdIndexTuple... (get output used as input in tx if it's spendable BSQ)
-        return readableBsqBlockChain.getUnspentAndMatureTxOutput(input.getTxIdIndexTuple());
-    }
+    void processInput(TxInput input, int blockHeight, String txId, int inputIndex, Model model,
+                      WritableBsqBlockChain writableBsqBlockChain) {
+        readableBsqBlockChain.getUnspentAndMatureTxOutput(input.getTxIdIndexTuple()).ifPresent(connectedTxOutput -> {
+            model.addToInputValue(connectedTxOutput.getValue());
+
+            // If we are spending an output from a blind vote tx marked as VOTE_STAKE_OUTPUT we save it in our model
+            // for later verification at the outputs of a reveal tx.
+            if (connectedTxOutput.getTxOutputType() == TxOutputType.BLIND_VOTE_STAKE_OUTPUT) {
+                if (!model.isVoteStakeSpentAtInputs()) {
+                    model.setVoteStakeSpentAtInputs(true);
+                } else {
+                    log.warn("We have a tx which has 2 connected txOutputs marked as VOTE_STAKE_OUTPUT. " +
+                            "This is not a valid BSQ tx.");
```

Comment:
The reveal tx need to take 1 stake output from the blind vote tx. there must not be 2 stake outputs. Just want to keep the rules as narrow as possible to avoid ambiguity and loopholes.

Reply:
For the reveal tx I agree. What happens if you don't reveal and want to transfer this as BSQ?

Reply:
Ah good point. I have not covered all the exceptional paths, but that will probably require some changes here. I prefer to first get the standard path solid and then tackle down all the exceptions to avoid that it gets too complex from the start...